File: stepik-union/src/main/java/org/stepik/plugin/actions/step/InsertStepikDirectives.java
package org.stepik.plugin.actions.step;

import com.intellij.icons.AllIcons;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.fileEditor.FileDocumentManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.stepik.core.StepikProjectManager;
import org.stepik.core.SupportedLanguages;
import org.stepik.core.courseFormat.StepNode;
import org.stepik.core.metrics.Metrics;
import org.stepik.core.utils.Utils;
import org.stepik.plugin.actions.ActionUtils;
import org.stepik.plugin.utils.ReformatUtils;

import static org.stepik.core.metrics.MetricsStatus.SUCCESSFUL;
import static org.stepik.core.utils.DirectivesUtilsKt.containsDirectives;
import static org.stepik.core.utils.DirectivesUtilsKt.getFileText;
import static org.stepik.core.utils.DirectivesUtilsKt.insertAmbientCode;
import static org.stepik.core.utils.DirectivesUtilsKt.removeAmbientCode;
import static org.stepik.core.utils.DirectivesUtilsKt.writeInToFile;
import static org.stepik.core.utils.ProjectFilesUtils.getOrCreateSrcDirectory;


public class InsertStepikDirectives extends CodeQuizAction {
    private static final String SHORTCUT = "ctrl alt pressed R";
    private static final String ACTION_ID = "STEPIK.InsertStepikDirectives";
    private static final String SHORTCUT_TEXT = ActionUtils.getShortcutText(SHORTCUT);
    private static final String TEXT = "Repair standard template (" + SHORTCUT_TEXT + ")";
    private static final String DESCRIPTION = "Insert Stepik directives. Repair ordinary template if it is possible.";

    public InsertStepikDirectives() {
        super(TEXT, DESCRIPTION, AllIcons.General.ExternalToolsSmall);
    }

    @NotNull
    @Override
    public String getActionId() {
        return ACTION_ID;
    }

    @Nullable
    @Override
    public String[] getShortcuts() {
        return new String[]{SHORTCUT};
    }

    @Override
    public void actionPerformed(AnActionEvent e) {
        Project project = e.getProject();
        StepNode stepNode = getCurrentCodeStepNode(project);
        if (stepNode == null) {
            return;
        }

        Utils.INSTANCE.saveAllDocuments(project);

        SupportedLanguages currentLang = stepNode.getCurrentLang();

        VirtualFile src = getOrCreateSrcDirectory(project, stepNode, true);
        if (src == null) {
            return;
        }

        VirtualFile file = src.findChild(currentLang.getMainFileName());
        if (file == null) {
            return;
        }

        String text = getFileText(file);

        StepikProjectManager projectManager = StepikProjectManager.getInstance(project);
        boolean showHint = projectManager != null && projectManager.getShowHint();
        boolean needInsert = !containsDirectives(text, currentLang);
        if (needInsert) {
            text = insertAmbientCode(text, currentLang, showHint);
            Metrics.insertAmbientCodeAction(project, stepNode, SUCCESSFUL);
        } else {
            text = removeAmbientCode(text, showHint, currentLang, true);
            Metrics.removeAmbientCodeAction(project, stepNode, SUCCESSFUL);
        }
        writeInToFile(text, file, project);
        if (needInsert) {
            final Document document = FileDocumentManager.getInstance().getDocument(file);
            if (document != null) {
                ReformatUtils.reformatSelectedEditor(project, document);
            }
        }

        currentLang.getRunner().updateRunConfiguration(project, stepNode);
    }
}

File: stepik-union/src/main/java/org/stepik/core/actions/step/InsertStepikDirectives.kt
package org.stepik.core.actions.step

import com.intellij.icons.AllIcons
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.fileEditor.FileDocumentManager
import org.stepik.core.StepikProjectManager
import org.stepik.core.actions.getShortcutText
import org.stepik.core.metrics.Metrics
import org.stepik.core.metrics.MetricsStatus.SUCCESSFUL
import org.stepik.core.utils.ProjectFilesUtils.getOrCreateSrcDirectory
import org.stepik.core.utils.Utils
import org.stepik.core.utils.containsDirectives
import org.stepik.core.utils.getFileText
import org.stepik.core.utils.insertAmbientCode
import org.stepik.core.utils.removeAmbientCode
import org.stepik.core.utils.writeInToFile
import org.stepik.plugin.utils.ReformatUtils


class InsertStepikDirectives : CodeQuizAction(TEXT, DESCRIPTION, AllIcons.General.ExternalToolsSmall) {

    override fun getActionId() = ACTION_ID

    override fun getShortcuts() = arrayOf(SHORTCUT)

    override fun actionPerformed(e: AnActionEvent) {
        val project = e.project
        val stepNode = getCurrentCodeStepNode(project) ?: return

        Utils.saveAllDocuments(project!!)

        val currentLang = stepNode.currentLang

        val src = getOrCreateSrcDirectory(project, stepNode, true) ?: return

        val file = src.findChild(currentLang.mainFileName) ?: return

        var text = getFileText(file)

        val projectManager = StepikProjectManager.getInstance(project)
        val showHint = projectManager != null && projectManager.showHint
        val needInsert = !containsDirectives(text, currentLang)
        if (needInsert) {
            text = insertAmbientCode(text, currentLang, showHint)
            Metrics.insertAmbientCodeAction(project, stepNode, SUCCESSFUL)
        } else {
            text = removeAmbientCode(text, showHint, currentLang, true)
            Metrics.removeAmbientCodeAction(project, stepNode, SUCCESSFUL)
        }
        writeInToFile(text, file, project)
        if (needInsert) {
            val document = FileDocumentManager.getInstance().getDocument(file)
            if (document != null) {
                ReformatUtils.reformatSelectedEditor(project, document)
            }
        }

        currentLang.runner.updateRunConfiguration(project, stepNode)
    }

    companion object {
        private const val SHORTCUT = "ctrl alt pressed R"
        private const val ACTION_ID = "STEPIK.InsertStepikDirectives"
        private val SHORTCUT_TEXT = getShortcutText(SHORTCUT)
        private val TEXT = "Repair standard template ($SHORTCUT_TEXT)"
        private const val DESCRIPTION = "Insert Stepik directives. Repair ordinary template if it is possible."
    }
}
